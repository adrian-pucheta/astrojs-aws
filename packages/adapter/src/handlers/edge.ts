import { Buffer } from "node:buffer"
import { URL } from "node:url"

import { polyfill } from "@astrojs/webapi"
import { App } from "astro/app"
import {
  CloudFrontHeaders,
  CloudFrontRequestHandler,
  CloudFrontRequestResult,
} from "aws-lambda"

polyfill(globalThis, {
  exclude: "window document",
})

export default function handler(
  app: App,
  knownBinaryMediaTypes: Set<string>,
): CloudFrontRequestHandler {
  return async function (event): Promise<CloudFrontRequestResult> {
    if (event.Records.length != 1) {
      throw new Error(
        "Illegal record size, cloudfront event records not equal 1",
      )
    }

    const { uri, method, headers, querystring, body } =
      event.Records[0].cf.request

    // convert aws cloudfront event to node request
    const requestHeaders = new Headers()
    for (const [key, values] of Object.entries(headers)) {
      for (const { value } of values) {
        if (value) {
          requestHeaders.append(key, value)
        }
      }
    }
    const host = (headers["x-forwarded-host"] || headers["host"])[0].value
    const qs = querystring.length > 0 ? `?${querystring}` : ""
    const url = new URL(`${uri}${qs}`, `https://${host}`)
    const request = new Request(url.toString(), {
      method,
      headers: requestHeaders,
      body: body?.data
        ? body.encoding === "base64"
          ? Buffer.from(body.data, "base64").toString()
          : body.data
        : undefined,
    })

    const routeData = app.match(request, { matchNotFound: true })
    if (!routeData) {
      return {
        status: "404",
        statusDescription: "Not found",
      }
    }

    // astro render
    const rendered = await app.render(request, routeData)

    // build cookies
    const responseHeaders: CloudFrontHeaders = {}
    const rawHeaders = rendered.headers.entries()
    for (const [key, value] of rawHeaders) {
      for (const v of value) {
        responseHeaders[key] = [
          ...(responseHeaders[key] || []),
          { key, value: v },
        ]
      }
    }

    // convert node response to cloudfront response
    const contentType = rendered.headers.get("content-type") ?? ""
    const responseIsBase64Encoded = knownBinaryMediaTypes.has(contentType)
    return {
      status: rendered.status.toString(),
      statusDescription: "OK",
      headers: responseHeaders,
      bodyEncoding: responseIsBase64Encoded ? "base64" : "text",
      body: await rendered.text(),
    }
  }
}
