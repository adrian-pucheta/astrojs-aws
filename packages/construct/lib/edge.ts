import {
  AllowedMethods,
  BehaviorOptions,
  Distribution,
  LambdaEdgeEventType,
} from "aws-cdk-lib/aws-cloudfront"
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs"
import { Bucket } from "aws-cdk-lib/aws-s3"
import { BucketDeployment, Source } from "aws-cdk-lib/aws-s3-deployment"
import { Construct } from "constructs"

import { AstroSiteConstruct } from "./construct"
import { EdgeAstroSiteProps } from "./types"

export class EdgeAstroSite extends AstroSiteConstruct {
  private readonly bucket?: Bucket
  private readonly function: NodejsFunction
  private readonly distribution?: Distribution
  readonly bucketArn: string
  readonly bucketName: string
  readonly functionArn: string
  readonly functionName: string
  readonly distributionId: string
  readonly domains: string[]

  constructor(scope: Construct, id: string, props: EdgeAstroSiteProps) {
    super(scope, id)

    const domains: string[] = []

    this.function = this.newFunction(
      this,
      props.serverEntry,
      props.serverOptions,
    )

    if (!(props.onlyLambda ?? false)) {
      this.bucket = this.newBucket(this, false)

      const defaultBehavior: BehaviorOptions = {
        origin: this.newS3Origin(this, this.bucket),
        allowedMethods: AllowedMethods.ALLOW_ALL,
        edgeLambdas: [
          {
            eventType: LambdaEdgeEventType.ORIGIN_REQUEST,
            functionVersion: this.function.currentVersion,
            includeBody: true,
          },
        ],
      }
      this.distribution = this.newDistribution(
        this,
        defaultBehavior,
        props.cfOptions,
      )

      new BucketDeployment(this, "BucketDeployment", {
        sources: [Source.asset(props.staticDir)],
        destinationBucket: this.bucket,
        distribution: this.distribution,
      })

      domains.push(`https://${this.distribution.distributionDomainName}`)
      domains.push(`https://${props.cfOptions?.domain ?? ""}`)
    }

    this.bucketArn = this.bucket?.bucketArn ?? ""
    this.bucketName = this.bucket?.bucketName ?? ""
    this.functionArn = this.function.functionArn
    this.functionName = this.function.functionName
    this.distributionId = this.distribution?.distributionId ?? ""
    this.domains = domains
  }
}
