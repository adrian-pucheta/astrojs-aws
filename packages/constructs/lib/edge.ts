import {
  AllowedMethods,
  Distribution,
  LambdaEdgeEventType,
} from "aws-cdk-lib/aws-cloudfront"
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs"
import { Bucket } from "aws-cdk-lib/aws-s3"
import { Construct } from "constructs"

import { AstroSiteConstruct } from "./construct"
import { EdgeAstroSiteProps } from "./types"

export class EdgeAstroSite extends AstroSiteConstruct {
  private readonly bucket: Bucket
  private readonly function: NodejsFunction
  private readonly distribution: Distribution
  readonly bucketArn: string
  readonly bucketName: string
  readonly distributionDomainName: string
  readonly domainName: string

  constructor(scope: Construct, id: string, props: EdgeAstroSiteProps) {
    super(scope, id)

    this.bucket = this.newBucket(this, props.staticDir, props.bucketOptions)
    this.function = this.newFunction(
      this,
      props.serverEntry,
      props.serverOptions,
    )

    this.function.grantInvokeUrl
    this.distribution = new Distribution(this, "Distribution", {
      ...props.distributionOptions,
      defaultBehavior: {
        ...props.distributionDefaultBehaviorOptions,
        origin: this.newS3Origin(this, this.bucket),
        allowedMethods:
          props.distributionDefaultBehaviorOptions?.allowedMethods ??
          AllowedMethods.ALLOW_ALL,
        edgeLambdas: [
          {
            eventType: LambdaEdgeEventType.ORIGIN_REQUEST,
            functionVersion: this.function.currentVersion,
            includeBody: true,
          },
          ...(props.distributionDefaultBehaviorOptions?.edgeLambdas ?? []),
        ],
      },
    })

    this.bucketArn = this.bucket.bucketArn
    this.bucketName = this.bucket.bucketName
    this.distributionDomainName = this.distribution.distributionDomainName
    this.domainName = this.distribution.domainName
  }
  /**
   * Get edge function
   *
   * @returns
   */
  GetFunction() {
    return this.function
  }
}
