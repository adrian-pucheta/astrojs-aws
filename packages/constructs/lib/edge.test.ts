import path from "path"
import { fileURLToPath } from "url"

import { Stack } from "aws-cdk-lib"
import { Template } from "aws-cdk-lib/assertions"
import { describe, expect, test } from "vitest"

import { EdgeAstroSite } from "./edge"

const __dirname = path.dirname(fileURLToPath(import.meta.url))

describe("Test LambdaAstroSite", () => {
  test("Expect match snapshot", () => {
    const stack = new Stack()
    new EdgeAstroSite(stack, "TestSite", {
      serverEntry: path.join(__dirname, "../tests/testdata/fake-entry.mjs"),
      staticDir: path.join(__dirname, "../tests/testdata/fakestatic"),
    })
    const template = Template.fromStack(stack)
    expect(template).toMatchSnapshot()
  })
})
