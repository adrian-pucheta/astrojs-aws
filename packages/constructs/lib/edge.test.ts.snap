// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test LambdaAstroSite > Expect match snapshot 1`] = `
{
  "Outputs": {
    "TestSiteBucketArn3086CE28": {
      "Value": {
        "Fn::GetAtt": [
          "TestSiteBucket722B9E9B",
          "Arn",
        ],
      },
    },
    "TestSiteBucketNameCA295A85": {
      "Value": {
        "Ref": "TestSiteBucket722B9E9B",
      },
    },
    "TestSiteCloudFrontDomainName28EBB168": {
      "Value": {
        "Fn::GetAtt": [
          "TestSiteDistribution7B61650D",
          "DomainName",
        ],
      },
    },
    "TestSiteFunctionArnE2BE3008": {
      "Value": {
        "Fn::GetAtt": [
          "TestSiteEdgeFunction0AA60362",
          "Arn",
        ],
      },
    },
    "TestSiteFunctionNameE92EE45D": {
      "Value": {
        "Ref": "TestSiteEdgeFunction0AA60362",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": {
      "DependsOn": [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "de91205450df70f943fbde8ad6f4eec685d4c82c11ddec668199a8d2ec4d7065.zip",
        },
        "Handler": "index.handler",
        "Layers": [
          {
            "Ref": "TestSiteBucketDeploymentAwsCliLayerD8E01993",
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TestSiteBucket722B9E9B",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "TestSiteBucket722B9E9B",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": [
          {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestSiteBucket722B9E9B": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "Tags": [
          {
            "Key": "aws-cdk:cr-owned:cc5bd9f8",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TestSiteBucketDeploymentAwsCliLayerD8E01993": {
      "Properties": {
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "a3b50bcf02563ef9ac5444957bb7982a8cc320ef6c48a2281481a3d997d34776.zip",
        },
        "Description": "/opt/awscli/aws",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "TestSiteBucketDeploymentCustomResource12F3C1EB": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DestinationBucketName": {
          "Ref": "TestSiteBucket722B9E9B",
        },
        "Prune": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": [
          {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
        ],
        "SourceObjectKeys": [
          "d392246897ac937e6e71a1323092f897e754affea87f754f3d252a22ceaca64d.zip",
        ],
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
    "TestSiteBucketPolicy38B317BC": {
      "Properties": {
        "Bucket": {
          "Ref": "TestSiteBucket722B9E9B",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Principal": {
                "CanonicalUser": {
                  "Fn::GetAtt": [
                    "TestSiteOriginAccessIdentity031C693B",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TestSiteBucket722B9E9B",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "TestSiteBucket722B9E9B",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "CanonicalUser": {
                  "Fn::GetAtt": [
                    "TestSiteOriginAccessIdentity031C693B",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "TestSiteBucket722B9E9B",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "TestSiteDistribution7B61650D": {
      "Properties": {
        "DistributionConfig": {
          "DefaultCacheBehavior": {
            "AllowedMethods": [
              "GET",
              "HEAD",
              "OPTIONS",
              "PUT",
              "PATCH",
              "POST",
              "DELETE",
            ],
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "LambdaFunctionAssociations": [
              {
                "EventType": "origin-request",
                "IncludeBody": true,
                "LambdaFunctionARN": {
                  "Ref": "TestSiteEdgeFunctionCurrentVersion108B3BE23f17ec91bf7d2addf37fc622de57e775",
                },
              },
            ],
            "OriginRequestPolicyId": "acba4595-bd28-49b8-b9fe-13317c0390fa",
            "ResponseHeadersPolicyId": "eaab4381-ed33-4a86-88ca-d9558dc6cd63",
            "TargetOriginId": "TestSiteDistributionOrigin1026CC2AE",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": [
            {
              "DomainName": {
                "Fn::GetAtt": [
                  "TestSiteBucket722B9E9B",
                  "RegionalDomainName",
                ],
              },
              "Id": "TestSiteDistributionOrigin1026CC2AE",
              "S3OriginConfig": {
                "OriginAccessIdentity": {
                  "Fn::Join": [
                    "",
                    [
                      "origin-access-identity/cloudfront/",
                      {
                        "Ref": "TestSiteOriginAccessIdentity031C693B",
                      },
                    ],
                  ],
                },
              },
            },
          ],
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "TestSiteEdgeFunction0AA60362": {
      "DependsOn": [
        "TestSiteEdgeFunctionServiceRole0A74CE15",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "5e1cfc71811b7b7ae660dfdcccc5570dee0bce2be7b1936a263ad125a5032ee6.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TestSiteEdgeFunctionServiceRole0A74CE15",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestSiteEdgeFunctionCurrentVersion108B3BE23f17ec91bf7d2addf37fc622de57e775": {
      "Properties": {
        "FunctionName": {
          "Ref": "TestSiteEdgeFunction0AA60362",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
    "TestSiteEdgeFunctionServiceRole0A74CE15": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "edgelambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestSiteOriginAccessIdentity031C693B": {
      "Properties": {
        "CloudFrontOriginAccessIdentityConfig": {
          "Comment": "Allows CloudFront to reach the bucket",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
